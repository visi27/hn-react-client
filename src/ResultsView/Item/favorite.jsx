import { PropTypes } from 'prop-types';
import React, { Component } from 'react';
import { Icon } from 'react-icons-kit';
import { heart } from 'react-icons-kit/fa/';
import ReactLoading from 'react-loading';

const IconHeart = props => <Icon icon={heart} {...props} />;
const Loading = <ReactLoading type="spin" color="#F4A261" height={18} width={18} />;

class Favorite extends Component {
  constructor(props) {
    super(props);

    const { itemId, isFavorite } = props;

    this.state = {
      isLoading: false,
      isFavorite: isFavorite(itemId),
    };
  }

  addToFavorites() {
    this.setState({ isLoading: true });
  }

  render() {
    const { itemId, onAddFavorite, onRemoveFavorite } = this.props;
    const { isLoading, isFavorite } = this.state;
    const FavIcon = isFavorite ? (
      <IconHeart
        style={{ color: '#F4A261' }}
        onClick={() => {
          this.setState({ isLoading: true });
          onRemoveFavorite(itemId);
          setTimeout(() => this.setState({ isLoading: false, isFavorite: false }), 1000);
          // this.forceUpdate();
        }}
      />
    ) : (
      <IconHeart
        onClick={() => {
          this.setState({ isLoading: true });
          onAddFavorite(itemId);
          setTimeout(() => this.setState({ isLoading: false, isFavorite: true }), 1000);
          // this.forceUpdate();
        }}
      />
    );

    return isLoading ? Loading : FavIcon;
  }
}

Favorite.propTypes = {
  itemId: PropTypes.string.isRequired,
  isFavorite: PropTypes.func,
  onAddFavorite: PropTypes.func,
  onRemoveFavorite: PropTypes.func,
};

Favorite.defaultProps = {
  isFavorite: () => {},
  onAddFavorite: () => {},
  onRemoveFavorite: () => {},
};

export default Favorite;
