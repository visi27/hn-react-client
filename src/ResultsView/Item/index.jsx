import moment from 'moment';
import { PropTypes } from 'prop-types';
import React, { Component } from 'react';
import { IconButton } from 'react-toolbox/lib/button';

class Item extends Component {
  render() {
    const {
      item, isFavorite, onAddFavorite, onRemoveFavorite,
    } = this.props;
    return (
      <div key={item.objectID} className="container-fluid item">
        <div className="list-inline">
          <span className="list-inline-item">
            <a className="h5 text-muted" href={item.url}>
              {item.title}
            </a>
          </span>
          <span className="list-inline-item">
            {isFavorite(item.objectID) ? (
              <IconButton
                icon="favorite"
                accent
                onClick={() => {
                  onRemoveFavorite(item.objectID);
                  this.forceUpdate();
                }}
              />
            ) : (
              <IconButton
                icon="favorite"
                onClick={() => {
                  onAddFavorite(item.objectID);
                  this.forceUpdate();
                }}
              />
            )}
          </span>
        </div>
        <div className="list-inline">
          <span className="list-inline-item">Author: {item.author}</span>
          <span className="list-inline-item">Created: {moment(item.created_at).fromNow()}</span>
          <span className="list-inline-item">Comments: {item.num_comments}</span>
          <span className="list-inline-item">Points: {item.points}</span>
          <span className="list-inline-item">
            <a className="text-muted" href={item.url}>
              ({item.url})
            </a>
          </span>
        </div>
        <hr />
      </div>
    );
  }
}

Item.propTypes = {
  item: PropTypes.shape({
    created_at: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    title_highlighted: PropTypes.string,
    url: PropTypes.string,
    author: PropTypes.string.isRequired,
    points: PropTypes.number.isRequired,
    story_text: PropTypes.string,
    num_comments: PropTypes.number.isRequired,
    objectID: PropTypes.string.isRequired,
  }).isRequired,
  isFavorite: PropTypes.func,
  onAddFavorite: PropTypes.func,
  onRemoveFavorite: PropTypes.func,
};

Item.defaultProps = {
  isFavorite: () => {},
  onAddFavorite: () => {},
  onRemoveFavorite: () => {},
};

export default Item;
