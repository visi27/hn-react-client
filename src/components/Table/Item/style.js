export const largeColumn = {
  width: '40%',
};
export const midColumn = {
  width: '30%',
};
export const smallColumn = {
  width: '10%',
};
