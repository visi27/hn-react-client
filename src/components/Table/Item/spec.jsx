import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Item from './index';

describe('Item', () => {
  const item = {
    title: '1',
    author: '1',
    num_comments: 1,
    points: 2,
    objectID: 'y',
    created_at: '2018-04-03T09:35:47.000Z',
  };

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Item item={item} />, div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(<Item item={item} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
