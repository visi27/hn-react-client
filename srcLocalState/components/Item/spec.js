import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Item from './index';

describe('Item', () => {
  const item = {
    title: '1',
    author: '1',
    num_comments: 1,
    points: 2,
    objectID: 'y'
  };

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Item item={item} />, div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(<Item item={item} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
