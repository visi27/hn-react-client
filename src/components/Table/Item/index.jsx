import React from 'react';
import { PropTypes } from 'prop-types';
import Button from '../../Button/presenter';
import * as style from './style';

const Item = ({ item, onDismiss }) => (
  <div key={item.objectID} className="container-fluid">
    <span style={style.largeColumn}>
      <a href={item.url}>{item.title}</a>
    </span>
    <div className="list-inline">
      <span className="list-inline-item">Author: {item.author}</span>
      <span className="list-inline-item">Comments: {item.num_comments}</span>
      <span className="list-inline-item">Points: {item.points}</span>
      <span className="list-inline-item">
        <Button className="btn btn-outline-danger btn-sm" onClick={() => onDismiss(item.objectID)}>
          Dismiss
        </Button>
      </span>
    </div>
  </div>
);

Item.propTypes = {
  item: PropTypes.shape({
    created_at: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    title_highlighted: PropTypes.string,
    url: PropTypes.string,
    author: PropTypes.string.isRequired,
    points: PropTypes.number.isRequired,
    story_text: PropTypes.string,
    num_comments: PropTypes.number.isRequired,
    objectID: PropTypes.string.isRequired,
  }).isRequired,
  onDismiss: PropTypes.func,
};

Item.defaultProps = {
  onDismiss: () => {},
};

export default Item;
