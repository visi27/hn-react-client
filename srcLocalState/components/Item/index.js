import React from 'react';
import Button from '../Button/';
import * as style from './style';

const Item = ({ item, onDismiss }) => (
  <div key={item.objectID} className="table-row">
    <span style={style.largeColumn}>
      <a href={item.url}>{item.title}</a>
    </span>
    <span style={style.midColumn}>{item.author}</span>
    <span style={style.smallColumn}>{item.num_comments}</span>
    <span style={style.smallColumn}>{item.points}</span>
    <span style={style.smallColumn}>
      <Button onClick={() => onDismiss(item.objectID)}>Dismiss</Button>
    </span>
  </div>
);

export default Item;
