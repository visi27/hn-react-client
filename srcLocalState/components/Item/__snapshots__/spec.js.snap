// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item has a valid snapshot 1`] = `
<div
  className="table-row"
>
  <span
    style={
      Object {
        "width": "40%",
      }
    }
  >
    <a
      href={undefined}
    >
      1
    </a>
  </span>
  <span
    style={
      Object {
        "width": "30%",
      }
    }
  >
    1
  </span>
  <span
    style={
      Object {
        "width": "10%",
      }
    }
  >
    1
  </span>
  <span
    style={
      Object {
        "width": "10%",
      }
    }
  >
    2
  </span>
  <span
    style={
      Object {
        "width": "10%",
      }
    }
  >
    <button
      className="button-inline"
      onClick={[Function]}
      type="button"
    >
      Dismiss
    </button>
  </span>
</div>
`;
